pageserver/src/consumption_metrics.rs

```diff
@@ -166,14 +169,34 @@ pub async fn collect_metrics_iteration(
             if timeline.is_active() {
                 let timeline_written_size = u64::from(timeline.get_last_record_lsn());
 
-                current_metrics.push((
-                    PageserverConsumptionMetricsKey {
-                        tenant_id,
-                        timeline_id: Some(timeline.timeline_id),
-                        metric: WRITTEN_SIZE,
-                    },
-                    timeline_written_size,
-                ));
+                let key = PageserverConsumptionMetricsKey {
+                    tenant_id,
+                    timeline_id: Some(timeline.timeline_id),
+                    metric: WRITTEN_SIZE,
+                };
+
+                // last_record_lsn can only go up, right now at least, TODO: #2592 or related
+                // features might change this.
+                //
+                // this will be None for the first item, do not send it then to avoid the need to
+                // filter out huge values from the stream of deltas.
+                let timeline_written_size_delta = cached_metrics
+                    .get(&key)
+                    .copied()
+                    .map(|prev| timeline_written_size - prev);
+
+                current_metrics.push((key, timeline_written_size));
+
+                if let Some(timeline_written_size_delta) = timeline_written_size_delta {
+                    current_metrics.push((
```

Comment:
BTW, I don't know this code well, but for deltas we need start / stop timestamps (there should be a relevent code on proxy side). Are they implicitly added by `push()` method?

- stop is just `now()`
- start is time of previous reporting OR pageserver start time (it's fine to have gaps)

Both could be the same for the entire events batch for all timelines

Reply:
Nothing is implicitly added here, timestamp is most likely read at the time of sending right now..

Reply:
Need to do a bit of refactoring for this.

Reply:
Refactorings at:
- [a4f3435](https://github.com/neondatabase/neon/pull/4822/commits/a4f3435182a948c0349cd966c17444405e0a0639)
- [674db26](https://github.com/neondatabase/neon/pull/4822/commits/674db26cd5b299cd5080f15ebc7296869d2c981e)
- [b620518](https://github.com/neondatabase/neon/pull/4822/commits/b6205180554db001fff246498ea0b0c28cb0d5d9)